### DISCUSSION TITLE
[Kavita] Add ability to read a comic by panel

### DISCUSSION THREAD

**Discussion by Gunnm92** [Kareadita/Kavita#2826] [Ideas]
### Idea Description

A like read comic by panel it's more confortable a better example for that is comic trim : https://play.google.com/store/apps/details?id=com.comic.trim&hl=en_US (the application is no longer available on the google store) 

On apk pure : 
https://apkpure.net/comic-trim/com.comic.trim/download

### Idea Category

User Experience

### Before submitting

- [X] I've already searched for existing ideas before posting.

**Comment by majora2007** [comment]
I looked at the linked app but not understanding the difference between our reader and what you're expecting.

**Reply by Gunnm92** [reply]
In this application, you can read a comic book panel by panel, which makes the experience much more enjoyable. There's no need to constantly zoom in, and it's more dynamic. It utilizes the panel's frame to identify it.

**Comment by DieselTech** [comment]
How do you tell what panel should come in which order? There is no metadata surrounding the panel.

**Reply by Gunnm92** [reply]
Maybe we can use it to give metadata of panel: 
https://framagit.org/nicooo/kumiko
https://github.com/CodeMinion/C.A.P.E
https://github.com/pedrovgs/DeepPanel

**Comment by Gunnm92** [comment]
I don't know how he coded the panel recognition, but it works very well in the application. Typically, we read from left to right or from right to left.

**Comment by mushroommunk** [comment]
I came to suggest the same thing as it only recently occurred to me. The link is dead for me, but it seems like the request is what is normally known as Guided View and for those of us who read mostly on our phones this would be fantastic. It would replace the current constant pinching to zoom and wiggle the image to see each box; not having to add yet another layer on top of Kavita would be even better in my book (I have too many apps around already).

It seems to me the best way to implement this would be to expand the [Anansi Comic Info Standard](https://github.com/anansi-project/comicinfo) that Kavita already supports (I don't know if @majora2007 has seen this come up already there or not, should I maybe create a parallel discussion there?). Expanding this metadata is something that I think should be done in general as many readers are haphazardly implementing things of that nature, might as well try to standardize it right? Either that or implement the [Advanced Comic Book Format](https://launchpad.net/acbf) which seems to already have this information alongside most the same things as ComicInfo, but that would obviously splinter things and seems to just duplicate work. It also doesn't make sense to me to create a second PanelInfo.xml or something just to hold this data.

Something like this would probably work (It's been a while since I've touched an xsd, probably needs tweaking):
`
<xs:complexType name="ComicPageInfo">
    <xs:attribute name="Image" type="xs:int" use="required"/>
    <xs:attribute default="Story" name="Type" type="ComicPageType"/>
    <xs:attribute default="false" name="DoublePage" type="xs:boolean"/>
    <xs:attribute default="0" name="ImageSize" type="xs:long"/>
    <xs:attribute default="" name="Key" type="xs:string"/>
    <xs:attribute default="" name="Bookmark" type="xs:string"/>
    <xs:attribute default="-1" name="ImageWidth" type="xs:int"/>
    <xs:attribute default="-1" name="ImageHeight" type="xs:int"/>
    <xs:element minOccurs="0" maxOccurs="1" name="Panels" type="ArrayOfComicPagePanels"/>
</xs:complexType>
<xs:complexType name="ArrayOfComicPagePanels">
    <xs:sequence>
        <xs:element minOccurs="0" maxOccurs="unbounded" name="Panel" nillable="true" type="ComicPagePanelInfo"/>
    </xs:sequence>
</xs:complexType>
<xs:simpleType name="ComicPagePanelInfo">
    <xs:attribute name="X" type="xs:int"/>
    <xs:attribute name="Y" type="xs:int"/>
    <xs:attribute name="W" type="xs:int"/>
    <xs:attribute name="H" type="xs:int"/>
</xs:simpleType>
`

Under ComicPagePanelInfo it might be prudent to support a series of points to define a more custom shape, but that does add complexity and adding later shouldn't break compatibility if done right.

Looking over manga-reader.component.ts it seems like it wouldn't be too painful (I know, famous last words) to then add a new display mode alongside Single and Double that scales and slides the image and adds black bars outside the specified box while looping through the panels the same way pages are looped through. It wouldn't matter then how the metadata got in the ComicInfo.xml file since any number of apps as Gunnm92 pointed out already can generate it. It looks like at least some of the metadata is already fetched and accessible in manga-reader component so don't think a server change would even be needed (though I haven't looked too close).

I'd even be happy to do the dev work if a consensus can be reached. Web apps in C# and Vue front end is my day job and I use Kavita a decent amount.

**Reply by majora2007** [reply]
Hi mushroommunk, I'd be open to a proof of concept for this guided view in a dedicated reader (like you see the webtoon reader is in a separate component). Showcasing the idea and how it might work would be interesting. 

I took a look at acbf, but I've never seen it in practice. I think on the fly processing might be an appropriate first step for a poc then identify if storing in the DB or in a dedicated xml/json would be beneficial during an analyze task. 

As for Anansi, I am a member already, but moving things through into the official standard has proven to be difficult without many members approvals, so I don't think that is an avenue. That is mainly around maintaining the ComicInfo.xml spec, which doesn't seem to be a good fit for something like this. 

I do think something like this would likely lead to a good experience on mobile, but there are quite many edge cases to consider and test against to see if it can be included into Kavita.

**Reply by mushroommunk** [reply]
Sure, I'll try and get something together over the next few days.

**Reply by majora2007** [reply]
No rush but excited to see what can be put together. Reach out if you need any help.

**Reply by mizaki** [reply]
I'll add this https://maxhalford.github.io/blog/comic-book-panel-segmentation/ for the splitting. And an implementation: https://sourceforge.net/projects/comic-panel-detection/ (that I've not tried yet).

As for external file formats. The problem I see with ACBF [frames](https://acbf.fandom.com/wiki/Body_Section_Definition?veaction=edit&section=16) is having to be in order. (Kumiko's JSON is also in order but there is a direction field.) 

To my mind having the order be split into columns and rows would enable easy reversing. That way it doesn't matter for the processing (oops I didn't pick the right direction) and left-to-right or right-to-left can be decided by the reader.

There is also [Digital Visual Narratives Profile](https://github.com/readium/webpub-manifest/blob/master/profiles/divina.md#4-guided-navigation) which, if I'm reading it right, another field could be added to the [media fragment](https://www.w3.org/TR/media-frags/) i.e. `image1.jpg#xywh=160,120,320,240&rowcol=1,3` or `image1.jpg#xywh=160,120,320,240&row=1&col=3` etc.

**Comment by mizaki** [comment]
After some more looking around I found https://github.com/Seeneva/seeneva-reader-android which uses ML to find the speech bubble and expands them one after another. So, it occurred that bubble pan/zoom might be more desirable as panels can be full page width etc.

A certain level of panel detection is still required (text below other text without panel knowledge could be the next text bubble in sequence (same panel) or 10 texts (and x panels) away) but I would imagine the "difficult" panels wouldn't be such a problem using text bubbles for guidance. The user being able to set a percentage zoom say for text bubbles.

I've found a few papers and githubs with examples but https://github.com/dmMaze/comic-text-detector which is used by https://github.com/kha-white/mokuro and https://github.com/dmMaze/BallonsTranslator seems to be used at least.

Making up a new format and trying to get similar people in form Anansi seems like a good idea (if the format moves fast enough)? Otherwise ACBF or Divina etc. Like information format, it makes sense (to me) that people can use whatever they want to generate the guiding boxes (panels or text).

I may try and wrap `comic-text-detector` to output an ACBF/Divina file.

**Comment by infectiousstupidity** [comment]
I like this idea. What about this project as inspiration? 

https://github.com/njean42/kumiko with a demo available at https://kumiko.njean.me/demo

**Comment by Gunnm92** [comment]
Hi I didn’t have much success with this idea, so I ended up doing it myself with an AI (I’m not a developer). It actually works quite well. In the end, the best approach is to use ML models for recognition, such as YOLO or U-Net, which perform very well. Max Halford’s method (without AI) also works, but it tends to “miss” some panels. It would require some fine-tuning, and that’s not something I know how to do.

**Comment by BillStrait** [comment]
Sorry to necromancer this thread, but it's a topic I've been thinking about lately. The concern here seems to be "magic up panel-to-panel navigation for every comic or manga." What if instead we check for the presence of a panel definition format (A future ComicInfo.xml, ACBF, or any/all of the other attempted formats out there) and offer panel-to-panel navigation? 

I think users would quickly identify which sources use a supported definition, but we can guide them along with a disabled button and infobox. This approach gets Kavita out of the role of co-authoring content.

**Reply by Gunnm92** [reply]
Contrary to what I thought, it's not really the panel segmentation that's the complex part — that can be done in code or by using vision models like YOLO, it's actually the simpler side. What's complex is everything around it: orchestrating the workers, the jobs, etc. If you want to maintain a decent user experience, it requires a lot of surrounding mechanisms. For example, I had to implement upscaling, because otherwise the cropped and enlarged panels are unreadable.

**Reply by BillStrait** [reply]
I'd suggest providing high res files suitable for zooming into the panels is also an authoring problem; but I'd be interested in seeing your approach. We are zooming in to those same images today without running into scaling issues, right?

Personally, this isn't work I want my software to do. It's a job I want the creators who designed the panels to do, or at least a human invested in interpreting their vision. I'm thinking about the "spiral" panels in a recent Marvel event, some "zig-zag" layouts in an indie book, and that no-panel Nightwing issue from a few years ago where no amount of coding will produce something useful. Automation will always be "best effort" and frequently wrong.

**Reply by Gunnm92** [reply]
I'd suggest providing high res files suitable for zooming into the panels is also an authoring problem; but I'd be interested in seeing your approach. We are zooming in to those same images today without running into scaling issues, right?

==> Unfortunately not it's not just a question of resolution but also of definition. 

Personally, this isn't work I want my software to do. It's a job I want the creators who designed the panels to do, or at least a human invested in interpreting their vision. I'm thinking about the "spiral" panels in a recent Marvel event, some "zig-zag" layouts in an indie book, and that no-panel Nightwing issue from a few years ago where no amount of coding will produce something useful. Automation will always be "best effort" and frequently wrong.

==> Initially I agreed with you, but it's not really a matter of artistic integrity (the scan parameters already corrupt the artist's work anywayn, if you truly want to respect the original, you need to go with physical paper). In my app I use RealCUGAN & Real-ESRGAN (one handles denoising, the other upscaling). The result is really very close to the original, with a sharpness that makes reading a pleasure. And it remains toggleable, by default it's an auto mode that decides whether to upscale based on resolution, but to be completely honest, I generally prefer the upscaled version, it's a more enjoyable read.


Not upscaled : 
<img width="1481" height="812" alt="Capture d’écran 2026-06-08 à 22 04 00" src="https://github.com/user-attachments/assets/95820616-872c-43b8-a0e0-ffd7cf466965" />


Upscaled : 
<img width="1479" height="796" alt="Capture d’écran 2026-06-08 à 22 04 24" src="https://github.com/user-attachments/assets/bcfa0a41-6e20-4e1b-a06d-96da6725e1ec" />

**Comment by burnshroom** [comment]
# Guided View proof of concept for Kavita

Following up on the Guided View proof of concept discussed here, I now have a working implementation integrated directly into Kavita.

It follows the direction suggested earlier in this discussion: **Kavita consumes, edits, and plays authored panel metadata without becoming responsible for automatic panel detection.**

## Demonstrations

These are informal, unedited screen captures of the prototype in use rather than narrated tutorials.

### Guided View in Kavita

This recording demonstrates the main Guided View reader and authoring workflow inside Kavita.

**[Watch the Guided View demonstration](https://youtu.be/eFgFIRA668)**

### Experimental Spoiler-Safe authoring

This recording demonstrates a separate experimental interface that obscures page details while panel regions are authored.

**[Watch the experimental Spoiler-Safe demonstration](https://youtu.be/nSP8fAZpdV8)**

> [!IMPORTANT]
> Spoiler-Safe authoring is shown for transparency about the wider exploration. It is **not part of the proposed V1–V2B contribution** and can be removed without affecting playback, normal authoring, metadata, or sidecars.

## What is implemented

| Area | Current behavior |
| --- | --- |
| **V1 — Playback** | Adds a panel-by-panel reader mode with forward/reverse navigation and normal page prefetching. |
| **Fallback** | Pages without panel metadata display normally, including books containing both authored and unauthored pages. |
| **V2A — Authoring** | Adds an admin-only editor for drawing, moving, resizing, deleting, and ordering panels. |
| **Import/export** | Supports importing, exporting, saving, and reverting Guided View definitions. |
| **V2B — Portability** | Writes a versioned JSON sidecar beside the archive without modifying the comic itself. |
| **Recovery** | A library scan can restore authored metadata from a matching sidecar after a database rebuild. |

The database remains Kavita's active operational copy. Portable metadata is stored beside the archive using its complete filename:

```text
Example.cbz
Example.cbz.guidedview.json
```

Invalid or mismatched sidecars are rejected rather than silently applied. If a library is read-only, the database edit is retained and the sidecar-write failure is reported as a warning.

<details>
<summary><strong>Example sidecar JSON</strong></summary>

```json
{
  "version": "1.0",
  "identity": {
    "fileName": "Example.cbz",
    "fileSize": 123456789,
    "koreaderHash": "optional-content-hash",
    "pageCount": 22
  },
  "pages": [
    {
      "image": "pages/001.jpg",
      "panels": [
        {
          "x": 0.03,
          "y": 0.04,
          "width": 0.94,
          "height": 0.28
        }
      ]
    }
  ]
}
```

The `image` value is the exact image-entry name inside the archive. Panel coordinates are normalized from `0.0` to `1.0`, making them independent of display resolution. Identity information helps prevent metadata from being applied to a different edition.

</details>

## Deliberately outside this proposal

- Automatic or ML-based panel detection
- Speech-bubble detection
- Community metadata hosting or curation
- Double-page/spread support currently being researched separately
- Experimental Spoiler-Safe authoring
- Modification of original comic archives

I separately tested classical computer-vision panel detection. It was fast for conventional bordered pages but unreliable across irregular, grouped, borderless, connected-border, and montage layouts, so that experiment was closed and is not included here.

The sidecar contract still leaves room for an independent tool to produce panel definitions later. Kavita would only need to validate, edit, and play that metadata.

<details>
<summary><strong>AI-use disclosure</strong></summary>

I am not a professional software developer. I explored and implemented this prototype using AI-assisted coding tools under my direction, including repository investigation, implementation, tests, debugging, and iteration.

I manually exercised and reviewed the resulting behavior. I am presenting this transparently as a working proof of concept, not claiming that AI-assisted code should be merged without normal maintainer scrutiny.

I understand that accepting a feature also means accepting its maintenance surface. I am willing to revise, separate, simplify, or remove implementation choices that do not fit Kavita's architecture or maintenance expectations.

</details>

## Proposed review approach

The complete V1–V2B implementation already exists in my fork. Because the combined change is fairly large, I propose reviewing it in two steps:

1. **V1 + V2A:** Guided View playback and manual authoring
2. **V2B:** Portable JSON sidecars as a follow-up

Would the maintainers be interested in reviewing a draft PR for the playback and manual-authoring foundation?

If a combined draft branch would be more useful for an initial architectural review, I can present that instead and reorganize it afterward based on maintainer feedback.

**Comment by aldo-o** [comment]
Depend on comic but i have not had too much luck with guided view detection so I pulled the coordinates from comics itself.
https://github.com/aldo-o/comic-panel-db

**Comment by Gunnm92** [comment]
"It was fast for conventional bordered pages but unreliable across irregular, grouped, borderless, connected-border, and montage layouts, so that experiment was closed and is not included here." ==> What vision model did you use? Because compared to what I did, it's the opposite: if you do detection with code, yes it's complicated on irregular panels, but with a vision model like YOLO or Mask R-CNN (especially the latter), it works very well even if it's irregular.
